### PR TITLE
(MODULES-6410) Rename distelli_agent to pipelines

### PR DESCRIPTION
**Pull request by danielparks**
This also makes a huge number of improvements and bug fixes. Notably:

  * Bumped the version to 1.0.0.
  * Updated module to follow the standard recommended by the [PDK][].
  * Added a trivial test that the manifest compiles.
  * Stopped managing chocolatey since the parameterization was more complicated than just requiring users to add `include chocolatey` to their manifest.
  * Unified code for Linux and macOS.
  * Made managing `/etc/sudoers` optional.
  * Moved the distelli user home to `/opt/distelli` on Linux and macOS.
  * Changed password default to `undef`. Linux and macOS do not require users to have passwords, and in fact prevent certain types of logins if a password is not present. Windows users now have to explicitly set a password.
  * Changed the user's default shell to `false` to prevent logins on Linux and macOS.
  * Made the user a "system" user on macOS and Linux.
  * Ensured that changes to the version parameter cause the agent to be upgraded to that version.
  * Added a prefix to certain internal resources names so they are less likely to be duplicated.
  * Stopped managing bzip and wget, since they weren't necessary.

Tested on Debian 8, CentOS 7, macOS 10.13, and Windows 2016.

[PDK]: https://puppet.com/docs/pdk/1.0/index.html

### CODE REVIEW

**Comment by danielparks** [comment]
Looks like this broke something on my build servers (it worked at first…). Investigating.

**Comment by danielparks** [comment]
It looks like the problem was with moving from one version of this module to another. Unfortunately, I can't tell exactly what change caused the issue. I had to uninstall the agent (`sudo /opt/distelli/distelli agent uninstall`) and then run puppet to get things back up and running. Further puppet runs are fine.

I've tested upgrading the `distelli` tool from 3.66.32 to 3.66.33, and it worked fine on Linux, macOS, and Windows.

**Review by mattkirby — APPROVED** [review]
Looks reasonable to me, I'm +1, but it's probably good to get at least one more review from someone with a little more familiarity with the code.

**Comment by danielparks** [comment]
Found a bug. Looks like pipelines may not play well if the `distelli` executable is outside of `/home/distelli`. I'll try to get to this later this week.

**Comment by danielparks** [comment]
Fixed. The issue was that the agent expects the `distelli` executable (a symlink) to be in the same directory as the versioned executable it points to.

File: README.md
Comment (glennsarti) [inline README.md]: Prefer not to use this kind of acronym.  Not every knows what this is.  Keep the "Beginning with..."

File: manifests/agent/params.pp
Comment (glennsarti) [inline manifests/agent/params.pp]: Dangerous default.  Doesn't have to be `c:\`, or `C:\users`, or even `C:\users\Distelli`

File: metadata.json
Comment (glennsarti) [inline metadata.json:26]: Err...Where's the Windows ❤️ ?  No OS Support entry for Windows.

File: manifests/agent/windows.pp
Comment (glennsarti) [inline manifests/agent/windows.pp]: Dangerous Default.  Do not hardcode this.

File: manifests/agent/windows.pp
Comment (glennsarti) [inline manifests/agent/windows.pp]: Dangerous Default.  Do not hardcode this

File: manifests/agent/windows.pp
Comment (glennsarti) [inline manifests/agent/windows.pp]: Dangerous Default. Do not hardcode system drive

Also, this is not the location for application data.  That should be  `%APPDATA%\PuppetLabs\pipelines` or similar.

**Review by glennsarti — CHANGES_REQUESTED** [review]
Too many dangerous defaults used.

File: manifests/agent/params.pp
Comment (danielparks) [reply manifests/agent/params.pp]: Sorry, I don't understand that. What should it be?

**Comment by danielparks** [comment]
@glennsarti thanks for the review. I don't know Windows at all, so it's particularly helpful. Could you tell me what the defaults should be?

File: manifests/agent/windows.pp
Comment (danielparks) [reply manifests/agent/windows.pp]: Application data: This has a similar problem on Linux; it installs the binary in `/home/distelli`. :( I reverted my change to make it `/opt/distelli` because every time something broke that was always the first thing people tried to revert, and it's a pain to change. 😭

Comment (Iristyle) [reply manifests/agent/windows.pp]: Does the Distelli agent itself expect this file to be here? If so, hands may be tied?

Comment (danielparks) [reply manifests/agent/windows.pp]: I'm not sure about Windows. On Linux and macOS, it works with a different location, but it's not the default when you use the installer.

Comment (danielparks) [reply manifests/agent/windows.pp]: I would love to make this follow common practice if not standards. I'm just running into practical problems, and trying to avoid them to get this merged.

If y'all think it's better to change the defaults to something more standard, I would be very happy to do that.

**Comment by danielparks** [comment]
I would love more reviews. If you know people who would be good reviewers please point them at this.

**Comment by Iristyle** [comment]
@danielparks the installation layout for the Puppet agent should be helpful when determining where to store certain data - https://github.com/puppetlabs/puppet-specifications/blob/master/file_paths.md

If you look around the agent upgrade module there will also be some clues about how to deal with some of the other comments Glenn made.

*  `TEMP` for instance is handled with a simple fact:
    * https://github.com/puppetlabs/puppetlabs-puppet_agent/blob/master/manifests/windows/install.pp#L37
    * https://github.com/puppetlabs/puppetlabs-puppet_agent/blob/f9e6b7743fbbd3a3a4df122f19f092c658c831f2/lib/facter/env_temp_variable.rb
* ProgramFiles could easily have a fact based off of ENV var

File: manifests/agent/params.pp
Comment (glennsarti) [reply manifests/agent/params.pp]: So the issue is the system drive doesn't have to be `C:\`.  Also, the Users directory changes depending on which OS you're on.  And the directory your user uses at logon can change (name collisions or corrupt profiles will cause a different home directory to be used).

However, this is all moot.  After digging around the module code, the only place $user_home is used is during the user creation.  Which you don't (and shouldn't _really_) enforce during user creation anyway.

i.e. remove
https://github.com/puppetlabs/puppetlabs-distelli_agent/blob/2c63a9db4c7ca83eb84a4be50c4427ae24e757d2/manifests/deps/windows.pp#L30

and

https://github.com/puppetlabs/puppetlabs-distelli_agent/blob/2c63a9db4c7ca83eb84a4be50c4427ae24e757d2/manifests/deps/windows.pp#L33

File: manifests/agent/windows.pp
Comment (glennsarti) [reply manifests/agent/windows.pp]: Should be using a TMP or TEMP fact

In this case I know that that temp directory is used during RDP sessions, not local logons.

File: manifests/agent/windows.pp
Comment (glennsarti) [reply manifests/agent/windows.pp]: SystemDrive doesn't have to `C:\`.  `Program Files` can change (32 v 64bit, localisation)

Instead use the `%PROGRAM FILES%` env var as a fact.

File: manifests/agent/windows.pp
Comment (glennsarti) [reply manifests/agent/windows.pp]: So in Windows there's a separation between application executables and application data (Program Files vs App Data)

From the distelli docs
https://pipelines.puppet.com/docs/agent/automated-install-of-the-distelli-agent/
```
Placing the File
The default location for placing the distelli.yml file is /etc.

On Windows, the default location is %SystemDrive%\distelli.yml.

You can override the default location with the -conf option.
```

IMO using %SYSTEMDRIVE% was a quick and dirty method, really should be using AppData (That's been around since WinXP/Server 2003 days)

**Comment by genebean** [comment]
@danielparks I'll see about adding some more basic tests to this too including ones for Windows and macOS so that its not just Linux being tested.

**Comment by genebean** [comment]
https://github.com/danielparks/puppetlabs-pipelines/pull/7 adds tests for macOS. As noted there, I ran into https://github.com/rodjek/rspec-puppet/issues/599 while trying to add Windows to the metadata and thereby the tests. @glennsarti or @Iristyle - any idea how to work around that bug or if a fix is in the works?

**Comment by genebean** [comment]
FYI - I turned on travis-ci for this repo so hopefully when the next commit / push is done on this PR it will trigger a build.

**Comment by genebean** [comment]
danielparks/puppetlabs-pipelines#8 should take care of the rubocop errors in the tests

File: metadata.json
Comment (Iristyle) [inline metadata.json]: Might be missing something, but it looks like some of the chocolatey integration was removed... is this dependency still needed?

File: manifests/agent/windows.pp
Comment (Iristyle) [inline manifests/agent/windows.pp]: Since the use of PowerShell here is so limited, I would recommend removing the PowerShell module dependency altogether and just stick to basic `exec` against `cmd`.  Can't this resource just be a `File` anyhow?

File: metadata.json
Comment (genebean) [reply metadata.json]: @Iristyle Based on the fourth bullet at the top I think so.

File: manifests/agent/windows.pp
Comment (Iristyle) [inline manifests/agent/windows.pp]: Same here on removing `powershell` dependency ... can just use a simpler `cmd` based `exec`

File: manifests/agent/windows.pp
Comment (genebean) [reply manifests/agent/windows.pp]: That would solve the windows testing issue too I mentioned earlier (I think)

File: metadata.json
Comment (danielparks) [reply metadata.json]: The module per se isn't dependent on it, but people may want to use chocolatey to manage things that the module is dependent on.

You can use the module without chocolatey, so I'll remove the dependency.

File: manifests/agent/windows.pp
Comment (danielparks) [reply manifests/agent/windows.pp]: Unfortunately I don't see a good way to change the configuration path from `%SystemDrive%\distelli.yml`, since it's baked into the distelli executable (I assume). I am changing it to use `%SystemDrive%`, though.

**Comment by danielparks** [comment]
I've made the requested changes, though I can't change the location of the configuration file to someplace more standard than the root of the system drive.

**Bug:** Puppet doesn't register "C:\\" as an absolute path name unless it's actually running on Windows, so rspec-puppet is currently failing when testing Windows code. (Running this on a Windows node works; I can run a trivial task on it.)

**Comment by Iristyle** [comment]
@danielparks you might get some `rspec-puppet` help from @rodjek or @DavidS

**Comment by rodjek** [comment]
@danielparks I see where the problem is in rspec-puppet and will create an issue for it. In the meantime, I believe you should be able to avoid the `absolute_path?` validation by passing `source` a URI instead (`file:///C:/ProgramFiles/Distelli/distelli`)

**Comment by danielparks** [comment]
`file:///c:…` fixed the absolute path problem, but now I have a new error:

~~~
  1) pipelines::agent on windows-2008 R2-x64 should compile into a catalogue without dependency cycles
     Failure/Error: it { is_expected.to compile }
     
     ArgumentError:
       wrong number of arguments (given 1, expected 0)
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/puppet-5.4.0/lib/puppet/provider/user/windows_adsi.rb:17:in `initialize'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/puppet-5.4.0/lib/puppet/provider/user/windows_adsi.rb:17:in `new'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/puppet-5.4.0/lib/puppet/provider/user/windows_adsi.rb:17:in `user'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/puppet-5.4.0/lib/puppet/provider/user/windows_adsi.rb:21:in `groups'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/puppet-5.4.0/lib/puppet/type/user.rb:332:in `retrieve'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/puppet-5.4.0/lib/puppet/property/list.rb:37:in `should'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/puppet-5.4.0/lib/puppet/type/user.rb:449:in `block (2 levels) in <module:Puppet>'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/puppet-5.4.0/lib/puppet/type.rb:2142:in `instance_eval'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/puppet-5.4.0/lib/puppet/type.rb:2142:in `block in autorelation'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/puppet-5.4.0/lib/puppet/type.rb:2081:in `block in eachautorequire'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/puppet-5.4.0/lib/puppet/type.rb:2080:in `each'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/puppet-5.4.0/lib/puppet/type.rb:2080:in `eachautorequire'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/puppet-5.4.0/lib/puppet/type.rb:2137:in `autorelation'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/puppet-5.4.0/lib/puppet/type.rb:2169:in `autorequire'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/puppet-5.4.0/lib/puppet/graph/relationship_graph.rb:172:in `block (2 levels) in build_autorelation_dependencies'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/puppet-5.4.0/lib/puppet/graph/relationship_graph.rb:171:in `each'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/puppet-5.4.0/lib/puppet/graph/relationship_graph.rb:171:in `block in build_autorelation_dependencies'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/puppet-5.4.0/lib/puppet/graph/relationship_graph.rb:170:in `each'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/puppet-5.4.0/lib/puppet/graph/relationship_graph.rb:170:in `build_autorelation_dependencies'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/puppet-5.4.0/lib/puppet/graph/relationship_graph.rb:25:in `populate_from'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/puppet-5.4.0/lib/puppet/resource/catalog.rb:262:in `relationship_graph'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-puppet-2.6.11/lib/rspec-puppet/matchers/compile.rb:142:in `cycles_found?'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-puppet-2.6.11/lib/rspec-puppet/matchers/compile.rb:25:in `matches?'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.7.0/lib/rspec/expectations/handler.rb:50:in `block in handle_matcher'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.7.0/lib/rspec/expectations/handler.rb:27:in `with_matcher'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.7.0/lib/rspec/expectations/handler.rb:48:in `handle_matcher'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.7.0/lib/rspec/expectations/expectation_target.rb:65:in `to'
     # ./spec/classes/agent_spec.rb:12:in `block (4 levels) in <top (required)>'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/example.rb:254:in `instance_exec'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/example.rb:254:in `block in run'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/example.rb:500:in `block in with_around_and_singleton_context_hooks'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/example.rb:457:in `block in with_around_example_hooks'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/hooks.rb:466:in `block in run'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/hooks.rb:604:in `run_around_example_hooks_for'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/hooks.rb:466:in `run'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/example.rb:457:in `with_around_example_hooks'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/example.rb:500:in `with_around_and_singleton_context_hooks'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/example.rb:251:in `run'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/example_group.rb:628:in `block in run_examples'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/example_group.rb:624:in `map'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/example_group.rb:624:in `run_examples'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/example_group.rb:590:in `run'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/example_group.rb:591:in `block in run'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/example_group.rb:591:in `map'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/example_group.rb:591:in `run'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/runner.rb:118:in `block (3 levels) in run_specs'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/runner.rb:118:in `map'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/runner.rb:118:in `block (2 levels) in run_specs'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/configuration.rb:1926:in `with_suite_hooks'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/runner.rb:113:in `block in run_specs'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/reporter.rb:79:in `report'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/runner.rb:112:in `run_specs'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/runner.rb:87:in `run'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/runner.rb:71:in `run'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/runner.rb:45:in `invoke'
     # /Users/daniel/.rbenv/versions/2.3.5/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/exe/rspec:4:in `<main>'
~~~

Weird.

**Comment by rodjek** [comment]
Yay, more bugs! ;)

**Comment by glennsarti** [comment]
Given how many commits and file changes there are here...may be time to split this out to separate PRs

**Comment by danielparks** [comment]
Unfortunately, this can't really be effectively split up. I've turned off the Windows tests, and I think this is ready for merge.

**Comment by danielparks** [comment]
@glennsarti could I get a re-review, please? @Iristyle, if you wanted to review too that would be rad.

File: metadata.json
Comment (glennsarti) [inline metadata.json]: Given this was manually test on 2016, should probably add it here

**Review by glennsarti — APPROVED** [review]

**Comment by genebean** [comment]
@danielparks I have a PR up that should fix the rubocop tests at https://github.com/danielparks/puppetlabs-pipelines/pull/10/files

**Comment by genebean** [comment]
Pinging @doct15 per HipChat conversation - all tests are now green

**Comment by genebean** [comment]
What else needs to be done to close this out?

**Comment by danielparks** [comment]
@brimworks please submit a PR on this branch rather than adding commits directly to it. It's too hard to review otherwise.

Also, I'd appreciate it if you split these up into logical chunks — the “WIP: Overhaul.” commit seems to contain a bunch of changes.

**Comment by brimworks** [comment]
@danielparks this was a mistake. I intended to push to my separate "overhaul" branch which is a branch from this branch (so my "tracking branch" wasn't appropriately updated).

Sorry,
-Brian

**Comment by brimworks** [comment]
Okay, all is fixed now! Sorry about pushing to the wrong branch.

**Comment by brimworks** [comment]
Okay, I've based all my latest changes off of your branch, and I've posted this PR for it (after resolving the failed merge):

https://github.com/puppetlabs/puppetlabs-distelli_agent/pull/6

I'm going to go ahead and close out this PR, since it will be implicitly merged in when the above PR is merged. Major thanks to all the contributors to this PR, comments are welcomed on the above PR.

**Pull request closed without merging**
